### ISSUE TITLE
为什么CPU的速度能那么快？

### ISSUE THREAD

**Issue by KeyKy** [Zehaos/MobileNet#22]
谢谢。

**Comment by Zehaos** [comment]
CPU的卷积实现通常会用到im2col（强烈建议看代码），也就是把feature map和卷积核排列成可以用矩阵相乘的形式，你会发现depthwise 1x1的卷积可以不用进行内存的重排，这是最关键的地方。然后通用矩阵乘法的CPU优化已经很到位了，并行程度很高，核越多按道理应该越快。我用的是4核的cpu，我想在移动端8核应该也不差。

**Comment by KeyKy** [comment]
@Zehaos 你的解释感觉Caffe已经实现了，这里是Caffe的源码
```
template <typename Dtype>
void BaseConvolutionLayer<Dtype>::forward_cpu_gemm(const Dtype* input,
    const Dtype* weights, Dtype* output, bool skip_im2col) {
  const Dtype* col_buff = input;
  if (!is_1x1_) {
    if (!skip_im2col) {
      conv_im2col_cpu(input, col_buffer_.mutable_cpu_data());
    }
    col_buff = col_buffer_.cpu_data();
  }
  for (int g = 0; g < group_; ++g) {
    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, conv_out_channels_ /
        group_, conv_out_spatial_dim_, kernel_dim_,
        (Dtype)1., weights + weight_offset_ * g, col_buff + col_offset_ * g,
        (Dtype)0., output + output_offset_ * g);
  }
}
```

1x1的的确没有调用im2col进行重排，按照你的解释，Caffe的实现应该和TF的速度一样快吧？但是我发现并不是这样。

**Comment by Zehaos** [comment]
@KeyKy 对，关键在于depthwise。 Caffe 没有depthwise conv这个layer，是用conv + group参数来实现的（具体实现我没看，有人说是for循环）。
有人已经写了这个layer, [click here](https://github.com/yonghenglh6/DepthwiseConvolution).
你可以试试，如果你能回馈下测试结果我会非常感激。

泽浩

**Comment by jy001227801** [comment]
@Zehaos CAFFE 是用循环来解决的 ，令 group_==channel 的个数，就是 MOBILE的对应通道相乘，而且感觉 ，你说的这个LAYER 没干什么事情 呢，调用的还是基类的卷积方法。

**Comment by KeyKy** [comment]
@jy001227801 这个LAYER 只写了CUDA的，CPU的实现貌似还是原来的Caffe conv+group

**Comment by jy001227801** [comment]
@KeyKy cuda的貌似是拿的tensor 的，直接做卷积，没有IM2COL的过程，CPU的运算，我感觉MOBILE是靠 AVX进行的加速。GPU是可以快。

**Comment by jy001227801** [comment]
@Zehaos  我也一直没想明白，CPU快在哪里，TF的Depth_conv的代码我也看了，类似于 CAFFE 的IM2COL的过程它也有，于CAFFE的区别就是访问的方式不一样，如果不开多线程或是什么的话，按理说应该和CAFFE  group_==channel 的速度是类似的吧？？求答疑解惑啊，困扰很久了。

**Comment by Zehaos** [comment]
@jy001227801 把代码链接贴出来一起研究下

**Comment by jy001227801** [comment]
@Zehaos   https://github.com/tensorflow/tensorflow/blob/359d6f9716c0bb9bd8201ce600da98b0481a8049/tensorflow/core/kernels/depthwise_conv_op.cc   里面的DepthwiseConv2DKernel  
https://github.com/tensorflow/tensorflow/blob/2d23960e29c3efb1b30c5cb6f8f3368635b97f29/tensorflow/core/kernels/depthwise_conv_op.h 里面的DepthwiseInputCopyOp ，
不知道我看的对不对。

**Comment by Zehaos** [comment]
@ccJia 刚看了下，里面有很多细节优化过，比如加pad有利与指令并行。
这个op优化过很多次，具体可以跟进这些历史看看。
[depthwise conv op history](https://github.com/tensorflow/tensorflow/commits/2d23960e29c3efb1b30c5cb6f8f3368635b97f29/tensorflow/core/kernels/depthwise_conv_op.h)

**Comment by KeyKy** [comment]
我的想法不知道对不对？我看了下源码，貌似TF的速度快 用了多线程 work_sharder + eigen?

**Comment by ccJia** [comment]
@KeyKy 是的,TF 是用了多线程，这个可以在编译时候，通过配置设定的。之前测试Eigen 和 Blas  没有特别大的速度差异。

**Comment by Zehaos** [comment]
@KeyKy @ccJia 不仅是多线程吧，[benchmark](https://github.com/tensorflow/tensorflow/commit/3f27cffdae7cf9b1404d664b28be42b933908b0a)里面给出的结果1个thread也很快。

**Comment by KeyKy** [comment]
@Zehaos @ccJia 你们有试过跑单线程的mobilenet速度是多少image/s？这个benchmark 2.7G items/s 还没看懂是什么意思

**Comment by Zehaos** [comment]
@KeyKy 给的指标是带宽

**Comment by ysh329** [comment]
鹿过……mark

**Comment by liu6381810** [comment]
@Zehaos 您好，我运行python tools/time_benchmark.py 得到的结果大概是0.04 htop看到的cpu利用率大概560%，但是当我用下面代码去测试mobilenet的时候得到的速度是0.07左右，两个的区别是什么？
`import tensorflow as tf

import mobilenet
import time


batch_size = 100
height = 224
width = 224
with tf.Session() as sess:
    inputs = tf.random_uniform((batch_size, height, width, 3))
    logits, _ = mobilenet.mobilenet(inputs)
    sess.run(tf.global_variables_initializer())
    a = time.time()
    output = sess.run(logits)
    print(time.time() - a)`

**Comment by Zehaos** [comment]
@liu6381810 你这段代码没有 warm up， time_benchmark.py里面是有的，而且跑了很多次算平均。

**Comment by liu6381810** [comment]
@Zehaos 谢谢回复！请问如何做warm up，还有warm up的作用会这么明显吗？刚才我跑了100次做平均时间没太大变化。

**Comment by Zehaos** [comment]
@liu6381810 就是先跑几个iter不算时间， warm up额。。。影响还是有的，因为执行指令都放到cache里了嘛，接下来cache的命中率就高。。。

**Comment by liu6381810** [comment]
@Zehaos 嗯好的谢谢

**Comment by MrAprils** [comment]
@Zehaos 想问下gpu的3ms是怎样的条件，我在TITANX上无法稳定复现3ms

**Comment by Zehaos** [comment]
@MrAprils 
My environment:
Ubuntu 16.04 LTS, Xeon E3-1231 v3, 4 Cores @ 3.40GHz, GTX1060, TF 1.0.1, CUDA8.0, CUDNN5.1

[Here](https://github.com/Zehaos/MobileNet/blob/master/tools/time_benchmark.py) is a script for time benchmarking.

**Comment by MrAprils** [comment]
2017-08-15 10:53:28.821269: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1052] Creating TensorFlow device (/gpu:0) -> (device: 0, name: TITAN Xp, pci bus id: 0000:02:00.0, compute capability: 6.1)
2017-08-15 10:53:31.591376: step 0, duration = 0.008
2017-08-15 10:53:31.655964: step 10, duration = 0.005
2017-08-15 10:53:31.715869: step 20, duration = 0.004
2017-08-15 10:53:31.795379: step 30, duration = 0.008
2017-08-15 10:53:31.873261: step 40, duration = 0.008
2017-08-15 10:53:31.951907: step 50, duration = 0.005
2017-08-15 10:53:32.015416: step 60, duration = 0.008
2017-08-15 10:53:32.075358: step 70, duration = 0.007
2017-08-15 10:53:32.132694: step 80, duration = 0.004
2017-08-15 10:53:32.194687: step 90, duration = 0.009
2017-08-15 10:53:32.279380: Forward across 100 steps, 0.007 +/- 0.002 sec / batch
这是在我的环境上跑的结果，增大num_batches后中间有一段时间为3ms,但不稳定，请问您说的3ms指的是最低还是均值

**Comment by Zehaos** [comment]
@MrAprils 
TITANX or TITANX Pascal?

**Comment by MrAprils** [comment]
TITAN Xp 和  TITANX Pascal 均尝试过

**Comment by Zehaos** [comment]
@MrAprils 
Average time.

**Comment by MrAprils** [comment]
2017-08-15 11:06:14.477316: step 890, duration = 0.006
2017-08-15 11:06:14.517504: step 900, duration = 0.005
2017-08-15 11:06:14.550230: step 910, duration = 0.003
2017-08-15 11:06:14.579801: step 920, duration = 0.003
2017-08-15 11:06:14.615377: step 930, duration = 0.005
2017-08-15 11:06:14.655590: step 940, duration = 0.003
2017-08-15 11:06:14.696793: step 950, duration = 0.005
2017-08-15 11:06:14.737471: step 960, duration = 0.004
2017-08-15 11:06:14.777763: step 970, duration = 0.003
2017-08-15 11:06:14.811013: step 980, duration = 0.003
2017-08-15 11:06:14.842362: step 990, duration = 0.003
2017-08-15 11:06:14.867415: Forward across 1000 steps, 0.004 +/- 0.001 sec / batch
这是 TITANX Pascal 1000次的结果，挺接近，有个问题 num_steps_burn_in 是为了去除刚开始不准确的时间吗？

**Comment by Zehaos** [comment]
@MrAprils YES.

**Comment by MrAprils** [comment]
@Zehaos 有尝试过将inception或者其他的网络中的卷积转换为depthwise这种结构吗，我试验的结果不太理想

**Comment by KeyKy** [comment]
@MrAprils xception

**Comment by BowieHsu** [comment]
@MrAprils shufflenet

**Comment by aureosun** [comment]
我照着您的方法用自己的建立的数据集训练了mobilenet模型，现在把这个模型用来做钱向，也就是做图片分类，输入为.jpg图片，输出为预测的每张图片的label,分类代码参考了tensorflow/tensorflow/examples/label_image/label_image.py 想问一下您这个怎么能够测出来输入的速度？用这个代码只能够对于单张图片做分类，目前我还不清楚怎么一个batch一个batch的输入图片然后一次得到一个batch 的labels?

**Comment by mollyStark** [comment]
@Zehaos 你好～我用time_benchmark.py在k40上测试，前向速度是0.029 +/- 0.003 sec / batch，请问3ms是怎么做到的啊？有加什么配置吗？

**Comment by yja1** [comment]
@Zehaos 
I want to test cpu time, but run with param --mode cpu or --mode gpu always output:
【TITAN X】
python time_benchmark.py


I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:193] kernel reported version is: 410.48.0
I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:300] kernel version seems to match DSO: 410.48.0
2019-09-04 15:38:25.818497: step 0, duration = 0.080
2019-09-04 15:38:26.487455: step 10, duration = 0.063
2019-09-04 15:38:27.143364: step 20, duration = 0.055
2019-09-04 15:38:27.789180: step 30, duration = 0.057
2019-09-04 15:38:28.415913: step 40, duration = 0.057
2019-09-04 15:38:29.073405: step 50, duration = 0.057
2019-09-04 15:38:29.740881: step 60, duration = 0.062
2019-09-04 15:38:30.423823: step 70, duration = 0.076
2019-09-04 15:38:31.069521: step 80, duration = 0.062
2019-09-04 15:38:31.736825: step 90, duration = 0.060
2019-09-04 15:38:32.354283: Forward across 100 steps, 0.066 +/- 0.006 sec / batch

-----------
python time_benchmark.py --model gpu

2019-09-04 15:51:29.038419: step 0, duration = 0.053
2019-09-04 15:51:29.674031: step 10, duration = 0.060
2019-09-04 15:51:30.355107: step 20, duration = 0.077
2019-09-04 15:51:30.991385: step 30, duration = 0.058
2019-09-04 15:51:31.632375: step 40, duration = 0.071
2019-09-04 15:51:32.320882: step 50, duration = 0.071
2019-09-04 15:51:32.963323: step 60, duration = 0.062
2019-09-04 15:51:33.628824: step 70, duration = 0.062
2019-09-04 15:51:34.271345: step 80, duration = 0.060
2019-09-04 15:51:34.912231: step 90, duration = 0.065
2019-09-04 15:51:35.436929: Forward across 100 steps, 0.064 +/- 0.006 sec / batch



cpu and gpu time same....
